File: frontend/rag/ingest_radiology_public.py
#!/usr/bin/env python3
"""
Public-source RAG ingestion (best-effort).

Goal:
- Download a small set of PUBLIC pages that are accessible without login/bot-challenges.
- Extract plain text + lightweight bullet points.
- Append as RAG chunks into frontend/rag/knowledge.txt using '---' separators.

Notes:
- Radiopaedia blocks unsanctioned scraping (Cloudflare challenge). This script does NOT attempt to bypass it.
- ACR RADS pages are often dynamic; this script records the URL and keeps only accessible text.
- For PDFs, reliable extraction requires a PDF text library; not included here to avoid extra deps.
"""

from __future__ import annotations

import html
import re
import sys
import time
import urllib.request
from dataclasses import dataclass
from pathlib import Path


KB_PATH = Path(__file__).resolve().parent / "knowledge.txt"


@dataclass(frozen=True)
class Source:
    tag: str
    url: str
    note: str = ""


SOURCES: list[Source] = [
    Source(tag="ESC Guidelines hub", url="https://www.escardio.org/Guidelines", note="hub page (links to all guidelines)"),
    Source(tag="ERS Guidelines hub", url="https://www.ersnet.org/guidelines/", note="filterable list of ERS/Joint docs"),
    Source(tag="EURETINA Guidelines", url="https://www.euretina.org/guidelines", note="guideline list page"),
    Source(tag="ESMO Guidelines hub", url="https://www.esmo.org/guidelines", note="may show limited content without login"),
    Source(tag="ACR RADS hub", url="https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems", note="often dynamic"),
    Source(tag="ACR Lung-RADS", url="https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems/Lung-RADS", note="often dynamic"),
    Source(tag="ACR BI-RADS", url="https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems/Bi-Rads", note="often dynamic"),
    Source(tag="ACR PI-RADS", url="https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems/PI-RADS", note="often dynamic"),
    Source(tag="ACR TI-RADS", url="https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems/TI-RADS", note="often dynamic"),
    Source(tag="ACR LI-RADS", url="https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems/LI-RADS", note="often dynamic"),
    Source(tag="ACR NI-RADS", url="https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems/NI-RADS", note="often dynamic"),
]


def fetch(url: str, timeout_s: int = 30) -> str:
    req = urllib.request.Request(
        url,
        headers={
            "User-Agent": "MedicinEvidence-RAG-Ingest/1.0 (+non-bot; public pages only)",
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        },
        method="GET",
    )
    with urllib.request.urlopen(req, timeout=timeout_s) as resp:
        raw = resp.read()
    # best-effort decode
    try:
        return raw.decode("utf-8", errors="replace")
    except Exception:
        return raw.decode(errors="replace")


def html_to_text(s: str) -> str:
    # remove scripts/styles
    s = re.sub(r"(?is)<(script|style|noscript).*?>.*?</\1>", " ", s)
    # replace <br> and block tags with newlines
    s = re.sub(r"(?i)<br\\s*/?>", "\n", s)
    s = re.sub(r"(?i)</(p|div|li|h1|h2|h3|h4|h5|tr)>", "\n", s)
    # strip tags
    s = re.sub(r"(?s)<[^>]+>", " ", s)
    s = html.unescape(s)
    # collapse whitespace
    s = re.sub(r"[ \t\r\f\v]+", " ", s)
    s = re.sub(r"\n\\s*\n\\s*\n+", "\n\n", s)
    return s.strip()


def summarize_lines(text: str, max_lines: int = 18) -> str:
    lines = [ln.strip() for ln in text.splitlines()]
    lines = [ln for ln in lines if ln and len(ln) >= 8]
    # remove obviously boilerplate lines
    drop = ("cookie", "privacy", "terms", "©", "copyright", "login", "subscribe")
    filtered: list[str] = []
    for ln in lines:
        low = ln.lower()
        if any(k in low for k in drop):
            continue
        if len(ln) > 220:
            ln = ln[:220].rstrip() + "…"
        filtered.append(ln)
        if len(filtered) >= max_lines:
            break
    return "\n".join(f"- {ln}" for ln in filtered)


def append_chunk(tag: str, url: str, note: str, bullets: str) -> None:
    ts = time.strftime("%Y-%m-%d")
    chunk = []
    chunk.append(f"[{tag}] (public ingest, {ts})")
    chunk.append(f"- URL: {url}")
    if note:
        chunk.append(f"- Note: {note}")
    if bullets.strip():
        chunk.append(bullets)
    else:
        chunk.append("- (No extractable text from this page via public ingest.)")
    chunk_txt = "\n".join(chunk).strip() + "\n"
    with KB_PATH.open("a", encoding="utf-8") as f:
        f.write("\n---\n\n")
        f.write(chunk_txt)


def main() -> int:
    if not KB_PATH.exists():
        print(f"ERROR: knowledge file not found at {KB_PATH}", file=sys.stderr)
        return 2
    ok = 0
    fail = 0
    for src in SOURCES:
        try:
            html_raw = fetch(src.url)
            text = html_to_text(html_raw)
            bullets = summarize_lines(text)
            append_chunk(src.tag, src.url, src.note, bullets)
            ok += 1
            print(f"OK: {src.tag}")
        except Exception as e:
            append_chunk(src.tag, src.url, src.note, f"- ERROR: {type(e).__name__}: {e}")
            fail += 1
            print(f"FAIL: {src.tag} ({e})", file=sys.stderr)
    print(f"Done. ok={ok} fail={fail}. Updated: {KB_PATH}")
    return 0 if ok else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend/rag/ingest_radiology_public.mjs
/**
 * Public-source Radiology RAG ingestion (Node.js, no deps).
 *
 * What it does:
 * - Fetches a list of PUBLIC pages that are typically accessible without login/bot challenges.
 * - Extracts visible text (best-effort) and summarizes into short bullet lines.
 * - Appends each page as a chunk to `frontend/rag/knowledge.txt` using `---` separators.
 *
 * What it does NOT do:
 * - It does not bypass bot protections (e.g., Radiopaedia Cloudflare challenge).
 * - It does not scrape paywalled/login-required guideline full text.
 * - It does not download/parse PDFs (keep KB short + avoid deps).
 *
 * Run:
 *   node frontend/rag/ingest_radiology_public.mjs
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const KB_PATH = path.join(__dirname, "knowledge.txt");

/** @type {{tag:string,url:string,note?:string}[]} */
const SOURCES = [
  { tag: "ESC Guidelines hub", url: "https://www.escardio.org/Guidelines", note: "hub page (links to all guidelines)" },
  { tag: "ERS Guidelines hub", url: "https://www.ersnet.org/guidelines/", note: "filterable list of ERS/Joint docs" },
  { tag: "EURETINA Guidelines", url: "https://www.euretina.org/guidelines", note: "guideline list page" },
  { tag: "ESMO Guidelines hub", url: "https://www.esmo.org/guidelines", note: "may show limited content without login" },
  { tag: "ACR RADS hub", url: "https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems", note: "often dynamic" },
  { tag: "ACR Lung-RADS", url: "https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems/Lung-RADS", note: "often dynamic" },
  { tag: "ACR BI-RADS", url: "https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems/Bi-Rads", note: "often dynamic" },
  { tag: "ACR PI-RADS", url: "https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems/PI-RADS", note: "often dynamic" },
  { tag: "ACR TI-RADS", url: "https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems/TI-RADS", note: "often dynamic" },
  { tag: "ACR LI-RADS", url: "https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems/LI-RADS", note: "often dynamic" },
  { tag: "ACR NI-RADS", url: "https://www.acr.org/Clinical-Resources/Clinical-Tools-and-Reference/Reporting-and-Data-Systems/NI-RADS", note: "often dynamic" },
  // Radiopaedia intentionally omitted from fetching (bot protection); keep only link index in KB.
];

function nowISODate() {
  const d = new Date();
  const pad = (n) => String(n).padStart(2, "0");
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`;
}

async function fetchText(url) {
  const res = await fetch(url, {
    method: "GET",
    headers: {
      "user-agent": "MedicinEvidence-RAG-Ingest/1.0 (public pages only)",
      accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    },
  });
  const body = await res.text();
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return body;
}

function htmlToText(html) {
  let s = String(html || "");
  // Remove scripts/styles/noscript
  s = s.replace(/<(script|style|noscript)[^>]*>[\s\S]*?<\/\1>/gi, " ");
  // Breaks for common block-ish tags
  s = s.replace(/<br\s*\/?>/gi, "\n");
  s = s.replace(/<\/(p|div|li|h1|h2|h3|h4|h5|tr|section|article)>/gi, "\n");
  // Strip tags
  s = s.replace(/<[^>]+>/g, " ");
  // Decode a few common entities
  s = s.replace(/&nbsp;/g, " ").replace(/&amp;/g, "&").replace(/&lt;/g, "<").replace(/&gt;/g, ">").replace(/&#39;/g, "'").replace(/&quot;/g, '"');
  // Collapse whitespace
  s = s.replace(/[ \t\r\f\v]+/g, " ");
  s = s.replace(/\n\s*\n\s*\n+/g, "\n\n");
  return s.trim();
}

function summarizeToBullets(text, maxLines = 18) {
  const rawLines = String(text || "").split("\n").map((l) => l.trim()).filter(Boolean);
  const drop = ["cookie", "privacy", "terms", "copyright", "©", "login", "subscribe"];
  const out = [];
  for (let ln of rawLines) {
    const low = ln.toLowerCase();
    if (drop.some((k) => low.includes(k))) continue;
    if (ln.length < 10) continue;
    if (ln.length > 220) ln = ln.slice(0, 220).trimEnd() + "…";
    out.push(`- ${ln}`);
    if (out.length >= maxLines) break;
  }
  return out.join("\n");
}

function appendChunk({ tag, url, note }, bullets) {
  const ts = nowISODate();
  const parts = [];
  parts.push(`[${tag}] (public ingest, ${ts})`);
  parts.push(`- URL: ${url}`);
  if (note) parts.push(`- Note: ${note}`);
  if (bullets && bullets.trim()) parts.push(bullets.trim());
  else parts.push(`- (No extractable text from this page via public ingest.)`);

  const chunk = `\n---\n\n${parts.join("\n").trim()}\n`;
  fs.appendFileSync(KB_PATH, chunk, "utf8");
}

async function main() {
  if (!fs.existsSync(KB_PATH)) {
    console.error(`ERROR: knowledge file not found at ${KB_PATH}`);
    process.exitCode = 2;
    return;
  }

  let ok = 0;
  let fail = 0;

  for (const src of SOURCES) {
    try {
      const html = await fetchText(src.url);

      // Detect common bot challenge page markers (don’t attempt bypass)
      const lower = html.toLowerCase();
      if (lower.includes("cloudflare") && lower.includes("challenge")) {
        throw new Error("Bot challenge page (cloudflare)");
      }

      const txt = htmlToText(html);
      const bullets = summarizeToBullets(txt);
      appendChunk(src, bullets);
      ok++;
      console.log(`OK: ${src.tag}`);
    } catch (e) {
      appendChunk(src, `- ERROR: ${String(e?.message || e)}`);
      fail++;
      console.error(`FAIL: ${src.tag} (${String(e?.message || e)})`);
    }
  }

  console.log(`Done. ok=${ok} fail=${fail}. Updated: ${KB_PATH}`);
}

await main();

